src/os/bluestore/BlueStore.cc

```diff
@@ -5363,706 +5225,466 @@ void BlueStore::_pad_zeros_tail(
+    bufferlist::iterator& blp,
+    WriteContext *wctx)
+{
+  dout(10) << __func__ << " 0x" << std::hex << offset << "~0x" << length
+	   << std::dec << dendl;
+  const uint64_t min_alloc_size = g_conf->bluestore_min_alloc_size;
+  assert(length < min_alloc_size);
+
+  bufferlist bl;
+  blp.copy(length, bl);
+
+  // use tail cache?  only worry about a perfect append for now.
+  // (fixme: we could also handle near-appends that fall within the
+  // same block or read_block or whatever.)
+  if (offset == o->onode.size &&
+      o->tail_bl.length() &&
+      offset == o->tail_offset + o->tail_bl.length()) {
+    // use tail?
+    bufferlist t = o->tail_bl;
+    t.claim_append(bl);
+    bl.swap(t);
+    offset -= o->tail_bl.length();
+    length += o->tail_bl.length();
+    dout(10) << __func__ << "  using tail_bl 0x" << std::hex
+	     << o->tail_offset << "~0x" << o->tail_bl.length()
+	     << ", write is now 0x" << offset << "~0x" << length
+	     << std::dec << dendl;
+  }
+
+  // look for an existing mutable blob we can use
+  bluestore_blob_t *b = 0;
+  map<uint64_t,bluestore_lextent_t>::iterator ep = o->onode.seek_lextent(offset);
+  if (ep != o->onode.extent_map.begin()) {
+    --ep;
+    b = c->get_blob_ptr(o, ep->second.blob);
+    if (ep->first + b->get_max_length() <= offset) {
+      ++ep;
     }
-  } else {
-    uint64_t head = 0;
-    uint64_t tail = 0;
-    if (offset % min_alloc_size) {
-      head = min_alloc_size - (offset % min_alloc_size);
-      offset += head;
-      if (length >= head)
-	length -= head;
+  }
+  while (ep != o->onode.extent_map.end()) {
+    if (ep->first >= offset + length) {
+      break;
     }
-    if ((offset + length) % min_alloc_size) {
-      tail = (offset + length) % min_alloc_size;
-      if (length >= tail)
-	length -= tail;
+    int64_t blob = ep->second.blob;
+    b = c->get_blob_ptr(o, ep->second.blob);
+    if (!b->is_mutable()) {
+      dout(20) << __func__ << " ignoring immutable " << blob << ": " << *b
+	       << dendl;
+      ++ep;
+      continue;
     }
-
-    dout(20) << "  initial full 0x" << std::hex << offset << "~0x" << length
-	     << ", head 0x" << head << " tail 0x" << tail << std::dec
-	     << dendl;
-
-    // include tail?
-    if (tail) {
-      if (allow_overlay && _can_overlay_write(o, tail)) {
-	dout(20) << "  tail 0x" << std::hex << tail << std::dec
-		 << " will be captured by overlay" << dendl;
-      } else {
-	bp = o->onode.find_lextent(orig_offset + orig_length - 1);
-	if (bp == o->onode.extent_map.end()) {
-	  dout(20) << "  tail 0x" << std::hex << tail << std::dec
-		   << " not yet allocated" << dendl;
-	  length += min_alloc_size;
-	} else if (bp->second.is_shared()) {
-	  dout(20) << "  tail shared" << dendl;
-	  length += min_alloc_size;
-	  shared_tail = true;
-	} else {
-	  dout(20) << "  tail 0x" << std::hex << tail << std::dec
-		   << " will presumably WAL" << dendl;
-	}
-      }
+    if (ep->first < ep->second.offset) {
```

Comment:
Probably I'm missing something but thus comparison looks strange - absolute offset vs. relative one.
Shouldn't it be something like
    if ((ep->first % min_alloc_unit) < ep->second.offset)